### ISSUE TITLE
TestBuilder Single-sided Buffer poor quality linework

### ISSUE THREAD

**Issue by skocibusic** [locationtech/jts#178]
Hi there,

we have a problem when creating a single sided buffer for a line string and a little digging led us to a problem with the creation of offset curves. Using the following geometry as an example:

Input:  `LINESTRING (6 8, 2 5, 6 5)`

JTS TestBuilder 1.14 will correctly create the offset curve with a distance of 1.

Result: `LINESTRING (6.6 7.2, 5 6, 6 6)`

Adding another point to the lower segment leads to quite a strange resulting offset curve.

Input:  `LINESTRING (6 8, 2 5, 4 5, 6 5)`
Result: `LINESTRING (6.6 7.2, 2.6 4.2, 2.5925925925925926 4.209876543209877, 2 5.987654320987654, 2 6, 6 6)`

What seems to happen is that both segments are just shifted and joined through a third segment creating a triangle.
This is not a problem for creating a regular buffer, but when creating a single sided buffer the resulting polygon covers the triangle area:

`POLYGON ((2.263374485596708 5.197530864197531, 2 5.987654320987654, 2 6, 3.3333333333333335 6, 6 8, 6.6 7.2, 5 6, 6 6, 6 5, 4 5, 3.6666666666666665 5, 2.6 4.2, 2.5925925925925926 4.209876543209877, 2.3292181069958846 5, 2 5, 2.263374485596708 5.197530864197531))`

This might be a duplicate of issue #25, but I am not able to verify that.

Kind reagrds,
Marko Skocibusic

**Comment by dr-jts** [comment]
One quick comment - this is not related to #25.  This issue is related to how offset curves are constructed, whereas the other one is probably due to robustness issues in buffer polygon construction (although hard to tell, since github has deleted the data attachment).

**Comment by dr-jts** [comment]
The issue here is that the offset curve generation originated with the two-side buffer algorithm.  There's some heuristics used to make sure that buffer curves produce accurate results.  As you note, for that construction the extra lines don't matter, since they are covered by the resulting buffer polygon.  

This must be a bad side-effect of one of the heuristics.  Unfortunately it's a bit of a science project to try and figure out a solution which does not impact the two-side buffer case.

**Comment by dr-jts** [comment]
For the record, here is the incorrect offset curve and the SSB result.

![image](https://user-images.githubusercontent.com/3529053/30613598-87af0a82-9d3d-11e7-9150-4e9795eed7ad.png)

![image](https://user-images.githubusercontent.com/3529053/30613616-99e300d2-9d3d-11e7-89f0-1c74365aa23f.png)

**Comment by dr-jts** [comment]
And it can get worse...  here is LINESTRING (200 70, 297 101, 324 51, 343 101, 440 70) with distance 100.

![image](https://user-images.githubusercontent.com/3529053/30613663-c9bcf9d4-9d3d-11e7-8001-6b67a5a499a0.png)

**Comment by dr-jts** [comment]
The problem is with the heuristics used to create offset linework which tracks around inside corners.  In general the overlap due to the inside corner can be very long, and it's hard to trim it while the curve is being computed.

**Comment by dr-jts** [comment]
One more comment - the SingleSidedBuffer functions in the TestBuilder were added for experimental and proof-of-concept purposes.  They aren't intended for production usage (as is evident from this issue!).

I have some new ideas about how to produce clean offset curves, and they might extend to improving SSB functionality as well.  Stay tuned for them to show up in the JTS Lab.

**Comment by dr-jts** [comment]
Here's a sneak preview:
![image](https://user-images.githubusercontent.com/3529053/30670846-e75a609a-9e18-11e7-84d6-bd4c7a8af784.png)

**Comment by skocibusic** [comment]
> The issue here is that the offset curve generation originated with the two-side buffer algorithm. There's some heuristics used to make sure that buffer curves produce accurate results. As you note, for that construction the extra lines don't matter, since they are covered by the resulting buffer polygon.
>
> This must be a bad side-effect of one of the heuristics. Unfortunately it's a bit of a science project to try and figure out a solution which does not impact the two-side buffer case.

I think this is not quite right. The offset curve generation affects the double sided buffer as well when using Mitre as join style. Please have a look at this example:

`LINESTRING (33413163.36756757 6024167.394594596, 33413164.400000002 6024161.2, 33413170.700949676 6024160.94651352, 33413171.476863 6024161.937958999, 33413173.516562365 6024163.646896194, 33413174.886529 6024164.794706001, 33413176.36098 6024165.900544, 33413177.374664 6024166.361308999, 33413179.217727 6024165.900544, 33413181.06079 6024164.518247, 33413181.598201 6024163.711844, 33413182.627393 6024163.412409, 33413184.193996 6024163.135950001, 33413185.668447 6024163.043796, 33413187.60832 6024162.828002, 33413187.961857 6024162.651233999, 33413189.552771 6024162.12093, 33413190.259844 6024160.883552, 33413190.46907465 6024160.151244123, 33413190.469074655 6024160.151244123, 33413199.200000003 6024159.800000002, 33413191.400000002 6024170.6, 33413176.366780605 6024173.156669965)`

Calculating the offset curve with distance 1 yields this result:
`LINESTRING (33413164.353961494 6024167.558993584, 33413165.25270968 6024162.166504483, 33413170.22921036 6024161.966300433, 33413170.689357795 6024162.5542669995, 33413170.834643163 6024162.7044795, 33413172.874342527 6024164.413416695, 33413174.24430916 6024165.561226503, 33413174.286529087 6024165.594706067, 33413175.76098009 6024166.700544066, 33413175.947177395 6024166.810910631, 33413176.960861396 6024167.2716756305, 33413177.173291765 6024167.340823789, 33413177.395772245 6024167.361086195, 33413177.617199253 6024167.331451592, 33413179.460262254 6024166.8706865925, 33413179.64749846 6024166.803481703, 33413179.817726932 6024166.700544052, 33413181.66078993 6024165.318247052, 33413181.78730004 6024165.205402851, 33413181.89293125 6024165.072810738, 33413182.22796355 6024164.570083678, 33413182.85463238 6024164.387759504, 33413184.31245074 6024164.13049777, 33413185.730825834 6024164.041848544, 33413185.77900634 6024164.0376655245, 33413187.718879342 6024163.821871525, 33413187.89160352 6024163.787038209, 33413188.05553259 6024163.722429695, 33413188.345504906 6024163.577443944, 33413189.868998405 6024163.069613418, 33413190.035900213 6024162.996479063, 33413190.18675979 6024162.894272234, 33413190.316567864 6024162.766386695, 33413190.421014182 6024162.617068864, 33413191.128087185 6024161.379690864, 33413191.221368007 6024161.158272919, 33413191.430598658 6024160.425965042, 33413191.418727994 6024160.422573425, 33413190.469074655 6024161.138898443, 33413190.46907465 6024161.151244123, 33413190.509272024 6024161.150435882, 33413197.18511783 6024160.881867374, 33413190.828913998 6024169.682764986, 33413176.19911993 6024172.170825201)`

And the buffer with distance 1 and join style Mitre and Mitre limit of 10:
`MULTIPOLYGON (((33413165.25270968 6024162.166504483, 33413170.22921036 6024161.966300433, 33413170.785809107 6024162.669988315, 33413172.874342527 6024164.413416695, 33413174.19182188 6024165.511522947, 33413175.94062682 6024166.823126336, 33413177.258121308 6024167.421986879, 33413179.63844381 6024166.826907226, 33413181.791679207 6024165.21198096, 33413182.22796355 6024164.570083678, 33413182.85463238 6024164.387759504, 33413184.31245074 6024164.13049777, 33413185.683167506 6024164.0491648195, 33413187.906168506 6024163.801875293, 33413188.345504906 6024163.577443944, 33413190.234531503 6024162.945220505, 33413190.979850836 6024161.640911401, 33413192.9065901 6024163.48802409, 33413197.18511783 6024160.881867374, 33413194.322531383 6024164.845448607, 33413195.78052168 6024166.24318437, 33413201.214967944 6024158.718566477, 33413189.674158487 6024159.182851911, 33413189.331563648 6024160.492478205, 33413188.874085024 6024161.293065953, 33413187.74118453 6024161.674555876, 33413187.32056959 6024161.853843441, 33413185.58192732 6024162.047252272, 33413184.054690827 6024162.137484228, 33413182.483240873 6024162.414798572, 33413180.957289975 6024162.858761507, 33413180.323688317 6024163.821073243, 33413178.779280014 6024164.979379261, 33413177.47072922 6024165.306516427, 33413176.87388873 6024165.035225564, 33413175.508090034 6024164.010876775, 33413174.158782203 6024162.880375692, 33413172.19961648 6024161.238912627, 33413171.190860137 6024159.92526938, 33413163.549154405 6024160.232694321, 33413162.381173644 6024167.230195608, 33413162.36805428 6024167.425790397, 33413162.393345643 6024167.620186348, 33413162.456075814 6024167.805912935, 33413162.5538341 6024167.975832789, 33413162.682863705 6024168.123415985, 33413162.838206097 6024168.242990985, 33413163.013891555 6024168.329962585, 33413163.203168582 6024168.38098852, 33413163.39876337 6024168.394107888, 33413163.59315932 6024168.368816522, 33413163.77888591 6024168.306086353, 33413163.948805764 6024168.208328067, 33413164.09638896 6024168.07929846, 33413164.215963956 6024167.923956068, 33413164.30293556 6024167.748270611, 33413164.353961494 6024167.558993584, 33413165.25270968 6024162.166504483)), 
  ((33413187.739681058 6024170.208144737, 33413176.19911993 6024172.170825201, 33413176.010012712 6024172.222476907, 33413175.834615882 6024172.310029128, 33413175.679669846 6024172.430117281, 33413175.551129088 6024172.578126447, 33413175.453933362 6024172.748368716, 33413175.39181784 6024172.934301772, 33413175.367169596 6024173.128780314, 33413175.38093584 6024173.324330639, 33413175.432587545 6024173.513437859, 33413175.52013977 6024173.688834689, 33413175.64022792 6024173.843780726, 33413175.788237087 6024173.972321482, 33413175.958479356 6024174.069517208, 33413176.144412413 6024174.131632727, 33413176.338890955 6024174.156280973, 33413176.53444128 6024174.142514729, 33413191.97301029 6024171.51512321, 33413192.993229035 6024170.102512645, 33413187.739681058 6024170.208144737)))`

Sorry I do not know how to provide a screenshot here.

**Comment by dr-jts** [comment]
Yes, there are issues with some of the two-sided buffer types as well - non-round endcaps and joins are particularly tricky.  Technically these will require a different approach to solve than for offset curves (which are in some ways simpler to compute).

Can you file a separate issue for the mitre join problem?

To add a picture, compute it in TestBuilder and then click File / Save Screen to Clipboard.  Then you can paste the image into the issue.  Here's the image for the above problem:

![image](https://user-images.githubusercontent.com/3529053/30753326-dc958e26-9f73-11e7-90ec-edae8fa1b2b7.png)

**Comment by dr-jts** [comment]
Turns out that the reason for the bad buffer curve generation is a very short line segment in the input:

`LINESTRING (3.341319046907465E7 6024160.151244123,  3.3413190469074655E7 6024160.151244123 )`

I think this should have been filtered out by the simplification pre-processing step. Will investigate further when I can

**Comment by skocibusic** [comment]
In #180 you mention:
> This is caused by a VERY short line segment in the input:
But isn't the problem here similar? When shifting the middle point in the lower line of my first example further to the right or remove it completely the output is valid.

This still yields the invalid result:
Input: `LINESTRING (6 8, 2 5, 4.9999 5, 6 5)`
But moving it just a tick further yields the valid result:
Input: `LINESTRING (6 8, 2 5, 5 5, 6 5)`
Result: 'LINESTRING (6.6 7.2, 5 6, 6 6)'
There seems to be a relation between the edge of the valid result (5 6) and the point in the lower segment. I hope this might be helpful.

**Comment by dr-jts** [comment]
@skocibusic Yes, at a high level - both of these issues are related to poor quality linework at joins.  But the exact details of the problems are different.

In #180 the problem is due to the mitre join being computed improperly on the very short segment.  But in that case buffering with a round join works fine.   A solution there is to remove the very short segment, which will not impact buffer line quality.  (It may also be possible to improve mitre join handling of short segments as well).

In this case, the issue is with how linework is generated for an inside corner, and how that interacts with single-sided buffer.  This is harder to fix - at the moment I don't have any good ideas.

**Comment by FObermaier** [comment]
>Here's a sneak preview of a new JTS offset curve algorithm

@dr-jts , can we have a look at the code?
